source/common/http/websocket/ws_handler_impl.cc

```diff
@@ -0,0 +1,143 @@
+#include "common/http/websocket/ws_handler_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/timer.h"
+#include "envoy/router/router.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/common/utility.h"
+#include "common/http/codes.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+#include "common/http/http1/codec_impl.h"
+#include "common/http/utility.h"
+#include "common/network/utility.h"
+
+#include "spdlog/spdlog.h"
+
+namespace Envoy {
+namespace Http {
+namespace WebSocket {
+
+WsHandlerImpl::WsHandlerImpl(Http::HeaderMap& request_headers,
+                             const Router::RouteEntry& route_entry, WsHandlerCallbacks& callbacks,
+                             Upstream::ClusterManager& cluster_manager)
+    : Filter::TcpProxy(nullptr, cluster_manager), request_headers_(request_headers),
+      route_entry_(route_entry), ws_callbacks_(callbacks) {}
+
+WsHandlerImpl::~WsHandlerImpl() {}
+
+void WsHandlerImpl::initializeUpstreamConnection(Network::ReadFilterCallbacks& callbacks) {
+  read_callbacks_ = &callbacks;
+  read_callbacks_->connection().addConnectionCallbacks(downstream_callbacks_);
+  const std::string& cluster_name = route_entry_.clusterName();
+  Upstream::ThreadLocalCluster* thread_local_cluster = cluster_manager_.get(cluster_name);
+
+  if (thread_local_cluster) {
+    ENVOY_CONN_LOG(debug, "creating connection to upstream cluster {}",
+                   read_callbacks_->connection(), cluster_name);
+  } else {
+    HeaderMapImpl headers{{Headers::get().Status, std::to_string(enumToInt(Http::Code::NotFound))}};
```

Comment:
I'm still not thrilled with how much code is copied here. Is it possible to refactor slightly so that the same init code runs, but if there is a failure during init you call an onInitFailure() virtual function that can do different stuff? I think that should work and be very clean and copy much less code.